app/Helpers/Docker.php

```diff
@@ -0,0 +1,163 @@
+    }
+
+    public function tagImages(Command $command, string $service = null, string $sourceTag, array $newTags): bool
+    {
+        if ($service) {
+            $services = [
+                $this->getImageUrlForService($command, $service),
+            ];
+        } else {
+            $services = $this->getImageUrlsForAllServices($command);
+        }
+
+        foreach ($services as $serviceImage) {
+            foreach ($newTags as $newTag) {
+
+                if ($newTag === $sourceTag) {
+                    // No point tagging the same thing
+                    continue;
+                }
+
+                try {
+                    $this->helpers->process()->withCommand([
+                        'docker', 'tag',
+                        $serviceImage . $sourceTag, $serviceImage . $newTag
+                    ])
+                    ->echoLineByLineOutput(true)
+                    ->run();
+
+                    $command->comment("Tagged " . $serviceImage . $newTag);
+                } catch (ProcessFailed $e) {
+                    $command->error("Unable to tag {$serviceImage}{$sourceTag} as {$serviceImage}{$newTag}");
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    public function determineTags(Command $command): array
+    {
+        // We'll start with the standard tag
+        $tags = [
+            $this->prefixedTag($command),
+        ];
+
+        if (
+            !$command->option('skip-additional-tags')
+        ) {
+            // Not skipping, let's add latest and the env
+            $tags[] = 'latest';
+
+            if ($environment = $command->option('environment')) {
+                $tags[] = $environment;
+            }
+        }
+
+        return $tags;
+    }
+
+    public function getImageUrlsForAllServices(Command $command): ?array
+    {
+        $config = $this->fetchComposeConfig($command);
+
+        if (!$config) {
+            return null;
+        }
+
+        return collect($config['services'])
+            ->transform(function ($serviceData, $serviceName) {
+                return $serviceName;
+            })
+            ->filter()
+            ->all();
+    }
+
+    public function getImageUrlForService(Command $command, $service): ?string
+    {
+        $config = $this->fetchComposeConfig($command);
+
+        if (!$config) {
+            return null;
+        }
+
+        return collect($config['services'])
+            ->filter(function ($serviceData, $serviceName) use ($service) {
+                return ($serviceName == $service);
+            })
+            ->transform(function ($serviceData, $serviceName) {
+                // We're only updating services that have an image attached
+                if (!isset($serviceData['image'])) {
+                    return null;
+                }
+
+                return $serviceData['image'];
+            })
+            ->first();
+    }
```

Comment:
I would say that these two could be combined, and then a large part of https://github.com/netsells/cli/blob/75d3813bbbad3447f38f839b7779877a7f1cf510/app/Commands/DeployEcsServiceUpdate.php#L212 could be changed to reuse this method here instead. Unless I misunderstood, but the current implementation of getImageUrlsForAllServices seems a bit fishy, as it returns service names rather than image urls.

```suggestion
    public function getImageUrlsForServices(Command $command, array $services = []): array
    {
        $config = $this->fetchComposeConfig($command);

        if (!$config) {
            return [];
        }
        
        return collect($config['services'])
            ->filter(function (array $serviceData, string $serviceName) use ($services) {
                return isset($serviceData['image']) && (empty($services) || in_array($serviceName, $services));
            })
            ->map(fn (array $serviceData) => $serviceData['image'])
            ->unique()
            ->values()
            ->all();
    }
```

Then `tagImages` could be updated to use:

```php
$services = $this->getImageUrlsForServices($command, $service ? [$service] : []);
```

And `gatherTargetImages` could be changed to:

```php
    protected function gatherTargetImages(): array
    {
        $imageUrls = $this->helpers->docker()->getImageUrlsForServices($this, $this->option('service'));

        return collect($imageUrls)
            ->map(fn (string $url) => explode(':', $url)[0])
            ->unique()
            ->all();
    }
```

Reply:
Yeah, I'm really not sure why I split it up. Will do some more testing and figure it out